url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+                            StringTokenizer parser;
+                            parser = new StringTokenizer(line);
+                            xSize = Integer.parseInt(parser.nextToken());
+                            ySize = Integer.parseInt(parser.nextToken());
+                            stage++;
+			}
+                        break;
+                    }
+
+                } while ( !exit );
+
+                retImage = new RGBImage();
+                retImage.initNoFill(xSize, ySize);
+                //byte barr[] = retImage.getRawImage();
+                //readBytes(bis, barr);
+
+                ByteBuffer bb = retImage.getRawImageDirectBuffer();
+                byte barr[] = new byte[xSize*3];
+                for ( int i = 0; i < ySize; i++ ) {
+                    readBytes(bis, barr);
+                    bb.put(barr);
+                }
+
+                //-------------------------------------------------------------
+                // Invert image
+                int x, y;
+                RGBPixel pa = new RGBPixel();
+                RGBPixel pb = new RGBPixel();
+
+                for ( y = 0; y < ySize/2; y++ ) {
+		    for ( x = 0; x < xSize; x++ ) {
+		        retImage.getPixelRgb(x, y, pa);
+		        retImage.getPixelRgb(x, ySize-y-1, pb);
+		        retImage.putPixelRgb(x, y, pb);
+		        retImage.putPixelRgb(x, ySize-y-1, pa);
+		    }
+		}
+                //-------------------------------------------------------------
+
+                bis.close();
+                fis.close();
+                return retImage;
+              }
+              catch ( Exception e ) {
+                  VSDK.reportMessage(null, VSDK.ERROR, "importRGB (B)",
+                                     "Cannot import image file \"" + inImageFd.getAbsolutePath() + "\"");
+		  e.printStackTrace();
+                 throw new ImageNotRecognizedException("Error reading internal file:\n" + e, inImageFd);
+            }
+        }
+        throw new ImageNotRecognizedException("Image not recognized", inImageFd);
+    }
+
+    private static void transferPixels(int[] ori, byte[] dest, int w, int h, int pixelDepth)
+    {
+        int bPos=0;
+        for(int i=0; i<w*h; i++)
+        {
+            dest[bPos]=(byte)(ori[i]>>16);
+            bPos++;
+            dest[bPos]=(byte)(ori[i]>>8);
+            bPos++;
+            dest[bPos]=(byte)(ori[i]>>0);
+            bPos++;
+            if(pixelDepth==32)
+            {
+                dest[bPos]=(byte)(ori[i]>>24);
+                bPos++;
+            }
+        }
+    }
+    
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary JPEG image format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */    
+    public static boolean exportJPG(File fd, Image img)
+    {
+        try {
+            FileOutputStream fos = new FileOutputStream(fd);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at line 311. The resource is closed at line 313. There are other execution paths that do not contain closure statements, for example, when ImagePersistence.exportJPG throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).